File: leetsync/1570-final-prices-with-a-special-discount-in-a-shop/final-prices-with-a-special-discount-in-a-shop.go
func finalPrices(prices []int) []int {
    stack:=make([]int, 0, len(prices))
    for i, discount:= range prices{
        for len(stack)>0 && discount<=prices[stack[len(stack)-1]]{
            prices[stack[len(stack)-1]]-=discount
            stack=stack[:len(stack)-1]
        }
        stack = append(stack, i)
    }
    return prices
}

File: leetsync/1570-final-prices-with-a-special-discount-in-a-shop/final-prices-with-a-special-discount-in-a-shop.java
class Solution {
    public int[] finalPrices(int[] prices) {
        int n = prices.length;
        Stack<Integer> stack = new Stack<>();
        for (int i = 0; i < n; i++) {
            while (!stack.isEmpty() && prices[i] <= prices[stack.peek()]) {
                prices[stack.pop()] -= prices[i];
            }
            stack.push(i);
        }
        return prices;
    }
}